lib/main.dart

```diff
@@ -11,115 +11,110 @@ class MyApp extends StatelessWidget {
+            'timestamp': DateTime.now().millisecondsSinceEpoch
+          },
+        );
+      });
+      textMessageController.clear();
+      scrollController.animateTo(
+        0.0,
+        duration: const Duration(milliseconds: 500),
+        curve: Curves.fastOutSlowIn,
+      );
+    }
   }
 
   @override
   Widget build(BuildContext context) {
-    // This method is rerun every time setState is called, for instance as done
-    // by the _incrementCounter method above.
-    //
-    // The Flutter framework has been optimized to make rerunning build methods
-    // fast, so that you can just rebuild anything that needs updating rather
-    // than having to individually change instances of widgets.
     return Scaffold(
       appBar: AppBar(
-        // TRY THIS: Try changing the color here to a specific color (to
-        // Colors.amber, perhaps?) and trigger a hot reload to see the AppBar
-        // change color while the other colors stay the same.
         backgroundColor: Theme.of(context).colorScheme.inversePrimary,
-        // Here we take the value from the MyHomePage object that was created by
-        // the App.build method, and use it to set our appbar title.
-        title: Text(widget.title),
+        title: const Text('Chat Page'),
       ),
-      body: Center(
-        // Center is a layout widget. It takes a single child and positions it
-        // in the middle of the parent.
+      body: Padding(
+        padding: const EdgeInsets.symmetric(horizontal: 16.0),
         child: Column(
-          // Column is also a layout widget. It takes a list of children and
-          // arranges them vertically. By default, it sizes itself to fit its
-          // children horizontally, and tries to be as tall as its parent.
-          //
-          // Column has various properties to control how it sizes itself and
-          // how it positions its children. Here we use mainAxisAlignment to
-          // center the children vertically; the main axis here is the vertical
-          // axis because Columns are vertical (the cross axis would be
-          // horizontal).
-          //
-          // TRY THIS: Invoke "debug painting" (choose the "Toggle Debug Paint"
-          // action in the IDE, or press "p" in the console), to see the
-          // wireframe for each widget.
-          mainAxisAlignment: MainAxisAlignment.center,
-          children: <Widget>[
-            const Text(
-              'You have pushed the button this many times:',
+          crossAxisAlignment: CrossAxisAlignment.start,
+          children: [
+            Expanded(
+              child: ListView.builder(
+                controller: scrollController,
+                itemCount: messages.length,
+                reverse: true,
+                padding: const EdgeInsets.symmetric(vertical: 8.0),
+                itemBuilder: (context, i) {
+                  var timestamp = DateTime.fromMillisecondsSinceEpoch(
```

Comment:
Use final.
Also, I would create a `ChatMessage` class. instead of parsing json(or map) from here.

```
class ChatMessage {
  DateTime time;
  string message;
  
  string get timestamp => Datetime.fromMillisecondsSinceEpoch();
}
```

When we implement the firebase, we can also implement the factory constructor in it for `fromJson()`

Reply:
Makes it difficult, no?

Reply:
It would be actually easier because the widget code would be less.
You don't need to explain the data class concept but just mention we do some business logic in there.

Reply:
Ok, I'll add but in a separate file or same?

Reply:
Just in the same file. ;)